File: DKImagePickerControllerDemo_OC/DKImagePickerControllerDemo_OC/Source/DKImagePickerController/DKPopoverViewController.m
//
//  DKPopoverViewController.m
//  DKImagePickerControllerDemo_OC
//
//  Created by 赵铭 on 2017/7/3.
//  Copyright © 2017年 zm. All rights reserved.
//

#import "DKPopoverViewController.h"

@interface DKPopoverView : UIView
@property (nonatomic, strong) UIView * contentView;
@property (nonatomic, assign) CGFloat arrowWidth;
@property (nonatomic, assign) CGFloat arrowHeight;
@property (nonatomic, strong) UIImageView * arrowImageView;
@end

@implementation DKPopoverView
- (instancetype)init{
    if (self = [super init]) {
        _arrowWidth = 20;
        _arrowHeight = 10;
        _arrowImageView = [UIImageView new];
        _arrowImageView.image = [self arrowImage];
        [self addSubview:_arrowImageView];
    }
    return self;
}

- (UIImage *)arrowImage{
    UIGraphicsBeginImageContextWithOptions(CGSizeMake(_arrowWidth, _arrowHeight), NO, [UIScreen mainScreen].scale);
    CGContextRef context = UIGraphicsGetCurrentContext();
    [[UIColor clearColor] setFill];
    CGContextFillRect(context, CGRectMake(0, 0, _arrowWidth, _arrowHeight));
    CGMutablePathRef arrowPath = CGPathCreateMutable();
    
    CGPathMoveToPoint(arrowPath, nil, _arrowWidth/2, 0);
    
    CGPoint points[] = {CGPointMake(_arrowWidth, _arrowHeight)};
    CGPathAddLines(arrowPath, nil, points, 1);
    
    CGPoint points1[] = {CGPointMake(0, _arrowHeight)};
    CGPathAddLines(arrowPath, nil, points1, 1);
    
    CGPathCloseSubpath(arrowPath);
    
    CGContextAddPath(context, arrowPath);
    
    CGContextSetFillColor(context, CGColorGetComponents([UIColor redColor].CGColor));
    CGContextDrawPath(context, kCGPathFill);
    UIImage * arrowImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();

    return arrowImage;
}

- (void)setContentView:(UIView *)contentView{
    if (_contentView != contentView) {
        _contentView = contentView;
        _contentView.layer.cornerRadius = 5;
        _contentView.clipsToBounds = YES;
        _contentView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
        [self addSubview:_contentView];

    }
    
}

- (void)layoutSubviews{
    [super layoutSubviews];
    self.arrowImageView.frame = CGRectMake((self.bounds.size.width - self.arrowWidth) / 2, 0, _arrowWidth, _arrowHeight);
    self.contentView.frame = CGRectMake(0, self.arrowHeight, self.bounds.size.width, self.bounds.size.height - _arrowHeight);
}








@end






@interface DKPopoverViewController ()
@property (nonatomic, strong) UIViewController * contentViewController;
@property (nonatomic, strong) UIView * fromView;
@property (nonatomic, strong) DKPopoverView * popoverView;
@end

@implementation DKPopoverViewController
+ (void)popoverViewController:(UIViewController *)viewController
                     fromView:(UIView *)fromView{
    UIWindow * window = [UIApplication sharedApplication].keyWindow;
    DKPopoverViewController * popoverViewController = [self new];
    popoverViewController.contentViewController = viewController;
    popoverViewController.fromView = fromView;
    
    [popoverViewController showInView:window];
    [window.rootViewController addChildViewController:popoverViewController];
}


+ (void)dismissPopoverViewController{
    UIWindow * window = [UIApplication sharedApplication].keyWindow;
    for (UIViewController * vc in window.rootViewController.childViewControllers) {
        if ([vc isKindOfClass:[DKPopoverViewController class]]) {
            [(DKPopoverViewController * )vc dismiss];
        }
    }
}
- (instancetype)init{
    if (self = [super init]) {
        _popoverView = [DKPopoverView new];
    }
    return self;
}
- (void)loadView{
    [super loadView];
    UIControl * backgroundView = [[UIControl alloc] initWithFrame:self.view.bounds];
    backgroundView.backgroundColor = [UIColor clearColor];
    [backgroundView addTarget:self action:@selector(dismiss) forControlEvents:UIControlEventTouchUpInside];
    backgroundView.autoresizingMask = self.view.autoresizingMask;
    self.view = backgroundView;
}
- (void)dismiss{
    [UIView animateWithDuration:0.2 animations:^{
        CGAffineTransform form = CGAffineTransformTranslate(self.popoverView.transform, 0, -(self.popoverView.bounds.size.height/2));
        self.popoverView.transform = CGAffineTransformScale(form, 0.01, 0.01);
        
    } completion:^(BOOL finished) {
        [self.view removeFromSuperview];
        [self removeFromParentViewController];
    }];
}
- (void)showInView:(UIView *)view{
    [view addSubview:self.view];
    self.popoverView.contentView = self.contentViewController.view;
    self.popoverView.frame = [self calculatePopoverViewFrame];
    
    CGAffineTransform form = CGAffineTransformTranslate(self.popoverView.transform, 0, -(self.popoverView.bounds.size.height/2));
    self.popoverView.transform = CGAffineTransformScale(form, 0.1, 0.1);
    [UIView animateWithDuration:0.6 delay:0 usingSpringWithDamping:0.7 initialSpringVelocity:1.3 options:UIViewAnimationOptionAllowUserInteraction animations:^{
        self.popoverView.transform = CGAffineTransformIdentity;
        self.view.backgroundColor = [UIColor colorWithWhite:0.4 alpha:0.4];
    } completion:^(BOOL finished) {
        
    }];
}

- (CGRect)calculatePopoverViewFrame{
    CGFloat popoverY = [self.fromView convertPoint:self.fromView.frame.origin toView:self.view].y + self.fromView.bounds.size.height;
    CGFloat popoverWidth = self.contentViewController.preferredContentSize.width;
    if (popoverWidth == UIViewNoIntrinsicMetric) {
        if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad) {
            popoverWidth = self.view.bounds.size.width * 0.6;
        }else{
            popoverWidth = self.view.bounds.size.width;
        }
    }
    CGFloat popoverHeight = MIN(self.contentViewController.preferredContentSize.height + self.popoverView.arrowHeight, self.view.bounds.size.height - popoverY - 40);
    
    return CGRectMake((self.view.bounds.size.width - popoverWidth)/2, popoverY, popoverWidth, popoverHeight);
    
}
- (void)viewDidLoad {
    [super viewDidLoad];
    [self.view addSubview:self.popoverView];
    // Do any additional setup after loading the view.
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: DKImagePickerController/DKPopoverViewController.swift
//
//  DKPopoverViewController.swift
//  DKImagePickerController
//
//  Created by ZhangAo on 15/6/27.
//  Copyright (c) 2015年 ZhangAo. All rights reserved.
//

import UIKit

open class DKPopoverViewController: UIViewController {
    
    open class func popoverViewController(_ viewController: UIViewController, fromView: UIView) {
        let window = UIApplication.shared.keyWindow!
        
        let popoverViewController = DKPopoverViewController()
        
        popoverViewController.contentViewController = viewController
        popoverViewController.fromView = fromView
        
        popoverViewController.showInView(window)
        window.rootViewController!.addChildViewController(popoverViewController)
    }
    
    open class func dismissPopoverViewController() {
        let window = UIApplication.shared.keyWindow!

        for vc in window.rootViewController!.childViewControllers {
            if vc is DKPopoverViewController {
                (vc as! DKPopoverViewController).dismiss()
            }
        }
    }
    
    private class DKPopoverView: UIView {
        
        var contentView: UIView! {
            didSet {
                contentView.layer.cornerRadius = 5
				contentView.clipsToBounds = true
                contentView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
                self.addSubview(contentView)
            }
        }
        
        let arrowWidth: CGFloat = 20
        let arrowHeight: CGFloat = 10
        fileprivate let arrowImageView: UIImageView = UIImageView()
        
        override init(frame: CGRect) {
            super.init(frame: frame)
            
            self.commonInit()
        }

        required init?(coder aDecoder: NSCoder) {
            super.init(coder: aDecoder)
            
            self.commonInit()
        }
        
        func commonInit() {
            arrowImageView.image = self.arrowImage()
            self.addSubview(arrowImageView)
        }
        
        override func layoutSubviews() {
            super.layoutSubviews()
            
            self.arrowImageView.frame = CGRect(x: (self.bounds.width - self.arrowWidth) / 2, y: 0, width: arrowWidth, height: arrowHeight)
            self.contentView.frame = CGRect(x: 0, y: self.arrowHeight, width: self.bounds.width, height: self.bounds.height - arrowHeight)
        }
        
        func arrowImage() -> UIImage {
            UIGraphicsBeginImageContextWithOptions(CGSize(width: arrowWidth, height: arrowHeight), false, UIScreen.main.scale)
            
            let context = UIGraphicsGetCurrentContext()
            UIColor.clear.setFill()
            context?.fill(CGRect(x: 0, y: 0, width: arrowWidth, height: arrowHeight))
            
            let arrowPath = CGMutablePath()
            
            arrowPath.move(to: CGPoint(x: arrowWidth / 2, y: 0))
            arrowPath.addLine(to: CGPoint(x: arrowWidth, y: arrowHeight))
            arrowPath.addLine(to: CGPoint(x: 0, y: arrowHeight))
            arrowPath.closeSubpath()

            context?.addPath(arrowPath)
            
            context?.setFillColor(UIColor.white.cgColor)
            context?.drawPath(using: CGPathDrawingMode.fill)

            let arrowImage = UIGraphicsGetImageFromCurrentImageContext()
            UIGraphicsEndImageContext()
            
            return arrowImage!
        }
    }
    
    var contentViewController: UIViewController!
    var fromView: UIView!
    private let popoverView = DKPopoverView()
    
    override open func loadView() {
        super.loadView()
        
        let backgroundView = UIControl(frame: self.view.frame)
        backgroundView.backgroundColor = UIColor.clear
        backgroundView.addTarget(self, action: #selector(dismiss as (Void) -> Void), for: .touchUpInside)
        backgroundView.autoresizingMask = self.view.autoresizingMask
        self.view = backgroundView
    }
    
    override open func viewDidLoad() {
        super.viewDidLoad()
        
        self.view.addSubview(popoverView)
    }

	@available(iOS, deprecated: 8.0)
    override open func didRotate(from fromInterfaceOrientation: UIInterfaceOrientation) {
		super.didRotate(from: fromInterfaceOrientation)
		
        UIView.animate(withDuration: 0.2, animations: {
            self.popoverView.frame = self.calculatePopoverViewFrame()
        })
    }
    
    func showInView(_ view: UIView) {
		view.addSubview(self.view)
		
		self.popoverView.contentView = self.contentViewController.view
        self.popoverView.frame = self.calculatePopoverViewFrame()
        
        self.contentViewController!.addObserver(self,
                                                forKeyPath: "preferredContentSize",
                                                options: .new,
                                                context: nil)
		
        self.popoverView.transform = self.popoverView.transform.translatedBy(x: 0, y: -(self.popoverView.bounds.height / 2)).scaledBy(x: 0.1, y: 0.1)
        UIView.animate(withDuration: 0.6, delay: 0, usingSpringWithDamping: 0.7, initialSpringVelocity: 1.3, options: .allowUserInteraction, animations: {
            self.popoverView.transform = CGAffineTransform.identity
            self.view.backgroundColor = UIColor(white: 0.4, alpha: 0.4)
        }, completion: nil)
    }
    
    func dismiss() {
        self.contentViewController.removeObserver(self, forKeyPath: "preferredContentSize")
        
        UIView.animate(withDuration: 0.2, animations: {
            self.popoverView.transform = self.popoverView.transform.translatedBy(x: 0, y: -(self.popoverView.bounds.height / 2)).scaledBy(x: 0.01, y: 0.01)
            self.view.backgroundColor = UIColor.clear
        }, completion: { result in
            self.view.removeFromSuperview()
            self.removeFromParentViewController()
        }) 
    }
	
	func calculatePopoverViewFrame() -> CGRect {
		let popoverY = self.fromView.convert(self.fromView.frame.origin, to: self.view).y + self.fromView.bounds.height

        let preferredContentSize = self.contentViewController.preferredContentSize
		var popoverWidth = preferredContentSize.width
		if popoverWidth == UIViewNoIntrinsicMetric {
			if UI_USER_INTERFACE_IDIOM() == .pad {
				popoverWidth = self.view.bounds.width * 0.6
			} else {
				popoverWidth = self.view.bounds.width
			}
		}
		
		let popoverHeight = min(preferredContentSize.height + self.popoverView.arrowHeight, view.bounds.height - popoverY - 40)
		
		return CGRect(
			x: (self.view.bounds.width - popoverWidth) / 2,
			y: popoverY,
			width: popoverWidth,
			height: popoverHeight
		)
	}
    
    open override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
        if keyPath == "preferredContentSize" {
            UIView.animate(withDuration: 0.2, animations: {
                self.popoverView.frame = self.calculatePopoverViewFrame()
            })
        } else {
            super.observeValue(forKeyPath: keyPath, of: object, change: change, context: context)
        }
    }
}
